### ISSUE TITLE
Feat: Asynchronous indexing for elasticsearch

### ISSUE THREAD

**Issue by grignards** [tracim/tracim#4149]
## Feature description and goals

The improved search (see #4097) will need to index a broad scope of data (for example the spaces names must be indexed for all contents). This indexing can be costly so we shouldn't do it during the HTTP API request handling.

- [x] delegate the indexing to the existing RQ worker, using a new queue name (`elasticsearch`?, `elasticsearch_index` ?)
- [x] trigger the indexing through the existing hooks (`on_user_created`/…)
- [x] the indexing should only be given the id of the entity which is to be indexed (as the indexed fields are broad a simple dictionary serialization wouldn't be enough and it is not possible to properly serialize a sqlalchemy object).
- [x] handle sync/async mode depending on `JOBS__PROCESSING_MODE` config param value.

## Prerequisites

check case where full/partial index is required:
- for user: https://github.com/tracim/tracim/issues/4095
- for space: https://github.com/tracim/tracim/issues/4134
- for content: https://github.com/tracim/tracim/issues/4146

## To be discussed before development 

- should this indexing mechanism be part of a plugin? (pros: enforces more separation of concerns, the functionality is optional, cons: implies changes in the configuration to enable elasticsearch)

**Comment by inkhey** [comment]
Note that there is 2 contexts where indexing currently occured:
- at create/update of content
- using tracimcli command line.

For the second one, it seems logical to keep the synchronous behavior.

**Comment by inkhey** [comment]
@grignards do we agree that async/sync for indexing rely on `jobs.processing_mode` parameter like for other async jobs usage. it seems the most logical way to handle this.

**Comment by inkhey** [comment]
> the indexing should only be given the id of the entity which is to be indexed (as the indexed fields are broad a simple dictionary serialization wouldn't be enough and it is not possible to properly serialize a sqlalchemy object).

We need also the type of entity unless you have separate indexing worker for each kind of entity, as we expect to index multiples kind of entity in different index.

**Comment by grignards** [comment]
> > the indexing should only be given the id of the entity which is to be indexed (as the indexed fields are broad a simple dictionary serialization wouldn't be enough and it is not possible to properly serialize a sqlalchemy object).
> 
> We need also the type of entity unless you have separate indexing worker for each kind of entity, as we expect to index multiples kind of entity in different index.

In my mind this would be part of the function called: `index_user(1)` for example.

**Comment by Skylsmoi** [comment]
@grignards @inkhey is this issue blocking or can it be done anytime during the 3.6 ?

**Comment by lebouquetin** [comment]
> should this indexing mechanism be part of a plugin? (pros: enforces more separation of concerns, the functionality is optional, cons: implies changes in the configuration to enable elasticsearch)

Why the plugin way do imply changes in the configuration?

According to the project deadline, changing working stuff is not the right moment: this will imply potential extra bugs, extra tests, potential extra fixes which are not expected.

**Comment by inkhey** [comment]
> @grignards @inkhey is this issue blocking or can it be done anytime during the 3.6 ?

It's not blocking (it can be done later) but seems to me more logical to do this at the beginning just after https://github.com/tracim/tracim/pull/4160, @grignards ?


> > should this indexing mechanism be part of a plugin? (pros: enforces more separation of concerns, the functionality is optional, cons: implies changes in the configuration to enable elasticsearch)
> 
> Why the plugin way do imply changes in the configuration?
> 
> According to the project deadline, changing working stuff is not the right moment: this will imply potential extra bugs, extra tests, potential extra fixes which are not expected.

Seems more logical to me to have all elasticsearch code in a separate "apps", should not been done now to avoid extra-work.

**Comment by grignards** [comment]
> It's not blocking (it can be done later) for now but seems to me more logical to do this at the beginning just after #4160, @grignards ?

Depends: we can say this is not needed for the 3.6 but not doing it 

> @grignards do we agree that async/sync for indexing rely on `jobs.processing_mode` parameter ? like for other async jobs usage. it seems the most logical way to handle this.

I partly agree: I'm ok to do that for coherency reasons, but in the mid-term I think that dropping the `sync` method would be simpler: it is never used in our docker/production and doesn't *really* simplify things for small setups without docker (as the jobs also need an external service for TLMs).

**Comment by inkhey** [comment]
This issue seems to me mostly ok:
- i do suggest: `elasticsearch_index` for the queue name.
- i do think this issue should only be about "put the work in an async queue", the specific full/partial index rules of here index should be solved in specific index ticket.

**Comment by lebouquetin** [comment]
>> @grignards do we agree that async/sync for indexing rely on jobs.processing_mode parameter ? like for other async jobs usage. it seems the most logical way to handle this.

> I partly agree: I'm ok to do that for coherency reasons, but in the mid-term I think that dropping the sync method would be simpler: it is never used in our docker/production and doesn't really simplify things for small setups without docker (as the jobs also need an external service for TLMs).

Actually it is done mannually when sysadmin want to reindex data. I think almost all systems including search engine (for example all django based web systems) are exploiting synchronous indexation process

**Comment by lebouquetin** [comment]
This issue is not clear on my point of view. Is it only about executing the update of the index async? In the description there is some information about `for example the spaces names must be indexed for all contents` which is another problem than "running the indexation async in order to avoid to make it in the HTTP process

**Comment by PhilippeAccorsi** [comment]
Done.
